File: scraping/bike.py
import requests
import re

from os import environ
from records import Database
from json import dumps
from json import loads


class Bike:
    def __init__(self):
        self.__host = environ['BIKE_HOST']
        self.__key = environ['BIKE_KEY']
        self.__db_host = environ['DB_HOST']
        self.__db_port = environ['DB_PORT']
        self.__db_user = environ['DB_USER']
        self.__db_passwd = environ['DB_PASSWD']
        self.__db_schema = environ['DB_SCHEMA']
        self.__db_name = environ['DB_NAME']
        self.__contracts_url = "{host}/contracts?apiKey={key}".format(
            host=self.__host, key=self.__key)
        self.__stations_url = "{host}/stations?apiKey={key}".format(
            host=self.__host, key=self.__key)
        self.__db_url = "{name}://{user}:{passwd}@{host}:{port}/{schema}".format(
            name=self.__db_name,
            user=self.__db_user,
            passwd=self.__db_passwd,
            host=self.__db_host,
            port=self.__db_port,
            schema=self.__db_schema)
        self.__db_conn = Database(self.__db_url)

    def save_stations(self):
        req = requests.get(self.__stations_url)

        if req.status_code != 200:
            return None

        data = loads(dumps(req.json()))
        sql = open('sql/INSERT_STATION.sql').read()
        sql = re.sub(r'\s+', r' ', sql)

        for station in data:
            query = sql.format(
                id=station['number'],
                contract_name=re.sub(r"'+", r'', station['contract_name']),
                name=re.sub(r"'+", r'', station['name']),
                pos_lat=station['position']['lat'],
                pos_lon=station['position']['lng'],
                banking=station['banking'],
                bonus=station['banking'],
                bike_stands=station['bike_stands'],
                available_bike_stands=station['available_bike_stands'],
                available_bikes=station['available_bikes'],
                status=station['status'],
                last_update=station['last_update'],
                address=re.sub(r"'+", r'', station['address']))

            self.__db_conn.query(query)

File: src/data-server/lib/bike.js
/**
 * This module is used to fetch real time bike data from JCD server
 * @module bike
 */

const process = require('process')
const https = require('https')
const queryString = require('querystring')
const env = require('./env')

const cache = {}

/**
 * The coordinate object
 * @typedef {Object} Coordinate
 * @property {number} lat
 * @property {number} lon
 */

/**
 * The JSON object return from JCD server
 * @typedef {Object} Station
 * @property {number} number
 * @property {string} contract_name
 * @property {string} name
 * @property {string} address
 * @property {Coordinate} position
 * @property {boolean} banking
 * @property {boolean} bonus
 * @property {string} status
 * @property {number} bike_stands
 * @property {number} available_bike_stands
 * @property {number} available_bike
 * @property {number} last_update
 */

/**
 * This function returns the current stations information
 * in the specified country
 * @function
 * @param {string} country
 * @returns {Promise.<Station>}
 */

const search = country => {
  const query = {
    contract: '',
    apiKey: env.env().bikeToken
  }

  const options = {
    hostname: 'api.jcdecaux.com',
    path: '/vls/v1/stations?',
    method: 'GET',
    headers: {
      Accept: 'application/json'
    }
  }

  return new Promise((resolve, reject) => {
    contracts().then(contracts => {
      let name = ''

      for (const contract of contracts) {
        if (
          contract.country_code !== null &&
          contract.country_code.toLowerCase() === country
        ) {
          name = contract.name
        }
      }

      query.contract = name
      options.path += queryString.stringify(query)

      const req = https.request(options, res => {
        if (res.statusCode !== 200) {
          reject({
            code: res.statusCode,
            message: res.statusMessage
          })
        }

        let data = ''

        res.on('data', chunk => {
          data += chunk
        })

        res.on('end', () => {
          resolve(geoJSON(JSON.parse(data)))
        })
      })

      req.on('error', err => {
        reject({
          code: 400,
          message: err
        })
      })

      req.end()
    })
  })
}

/**
 * The Contract type
 * @typedef {Object} Contract
 * @property {string} name
 * @property {string} commerical_name
 * @property {string} country_code
 * @property {Array.<string>} cities
 */

/**
 * This function returns a list of contracts
 * @return {Contract}
 */

const contracts = () => {
  const query = {
    apiKey: env.env().bikeToken
  }

  const options = {
    hostname: 'api.jcdecaux.com',
    path: '/vls/v1/contracts?' + queryString.stringify(query),
    method: 'GET',
    headers: {
      Accept: 'application/json'
    }
  }

  return new Promise((resolve, reject) => {
    const req = https.request(options, res => {
      if (res.statusCode !== 200) {
        reject({
          code: res.statusCode,
          message: res.statusMessage
        })
      }

      let data = ''

      res.on('data', chunk => {
        data += chunk
      })

      res.on('end', () => {
        resolve(JSON.parse(data))
      })
    })

    req.on('error', err => {
      reject({
        code: 400,
        message: err
      })
    })

    req.end()
  })
}

/**
 * The GeoJSON data type
 * @typedef {Object} GeoJSON
 * @property {string} type
 * @property {Feature} features
 */

/**
 * The Feature data type
 * @typedef {Object} Feature
 * @property {string} type
 * @property {number[]} bbox
 * @property {Object} geometry
 * @property {string} geometry.type
 * @property {number[]} geometry.coordinates
 */

/**
 * Generate GeoJSON
 * @param {Station[]} stations - The data return from JCD
 * @returns {GeoJSON}
 */

const geoJSON = stations => {
  let data = {
    type: 'FeatureCollection',
    features: []
  }

  for (const station of stations) {
    data.features.push({
      type: 'feature',
      geometry: {
        type: 'Point',
        coordinates: [station.position.lat, station.position.lng]
      },
      properties: {
        banking: station.banking,
        bonus: station.bonus,
        bike_stands: station.bike_stands,
        available_bike_stands: station.available_bike_stands,
        available_bikes: station.available_bikes,
        status: station.status
      }
    })
  }

  return data
}

module.exports = {
  search: search,
  geoJSON: geoJSON,
  cache: cache
}
